gflownet/utils/batch.py

```diff
@@ -0,0 +1,346 @@
+                    self.all_possible_parents.append(parents)
+                    self.all_possible_parents_actions.append(parents_a)
+                if self.loss == "trajectorybalance":
+                    self.masks_invalid_actions_backward.append(
+                        env.get_mask_invalid_actions_backward(
+                            env.state, env.done, [action]
+                        )
+                    )
+            else:
+                if env.done:
+                    self.states.append(env.state)
+                    self.env_ids.append(env.id)
+                    self.steps.append(env.n_actions)
+
+    def process_batch(self):
+        """
+        Process internal lists to more convenient formats:
+        - converts and stacks list to a single torch tensor
+        - computes trajectory indicies (indecies of the states in self.states corresponding to each trajectory)
+        - if needed, computes states and parents in policy formats (stored in self.states_policy, self.parents_policy)
+        """
+        self.env_ids = tlong(self.env_ids, device=self.device)
+        self._process_states()
+        self.steps = tlong(self.steps, device=self.device)
+        self._process_trajectory_indices()
+        # process other variables, if we are in the train mode and recorded them
+        if len(self.actions) > 0:
+            self.actions = tfloat(
+                self.actions, device=self.device, float_type=self.float
+            )
+            self.done = tbool(self.done, device=self.device)
+            self.masks_invalid_actions_forward = tbool(
+                self.masks_invalid_actions_forward, device=self.device
+            )
+            if self.loss == "flowmatch":
+                self.all_possible_parents_state_idx = tlong(
+                    sum(
+                        [
+                            [idx] * len(p)
+                            for idx, p in enumerate(self.all_possible_parents)
+                        ],
+                        [],
+                    ),
+                    device=self.device,
+                )
+                self.all_possible_parents_actions = torch.cat(
+                    [
+                        tfloat(x, device=self.device, float_type=self.float)
+                        for x in self.all_possible_parents_actions
+                    ]
+                )
+            elif self.loss == "trajectorybalance":
+                self.masks_invalid_actions_backward = tbool(
+                    self.masks_invalid_actions_backward, device=self.device
+                )
+            self._process_parents()
+        self.is_processed = True
+
+    def _process_states(self):
+        """
+        Converts self.statees from list to torch tensor, computes states in the policy format
+        (stored in self.states_policy)
+        """
+        self.states = tfloat(self.states, device=self.device, float_type=self.float)
+        self.states_policy = self.states2policy()
```

Comment:
In order to make the API more readable and transparent, I would make this method return `states` and `states_policy` (instead of updating/creating `self.xxx`). What do you think?

Reply:
I am having second thoughts though, since the the method is private (`_method`). Interested in your opinion too @carriepl

Reply:
Done [here](https://github.com/alexhernandezgarcia/gflownet/pull/111/commits/d3631459ec9ae11edfb1253cc58fd0d6eb71aef5)